src/s3/util/fileViewerManager.ts

```diff
@@ -78,86 +90,146 @@ export class S3FileViewerManager {
+        }
+
+        let type
+        let charset
+        if (headResponse) {
+            type = mime.contentType(headResponse.ContentType!)
+            charset = mime.charset(type as string)
+        } else {
+            type = ''
+            charset = ''
+        }
+
+        return charset ? charset : ''
     }
 
     /**
      * Given an S3FileNode, this function:
      * Checks and creates a cache to store downloads
      * Retrieves previously cached files on cache and
      * Downloads file from S3 ands stores in cache
-     * Opens the tab on read-only with the use of an S3Tab
+     * Opens the tab on read-only with the use of an S3Tab, or shifts focus to an edit tab if any.
      *
      * @param fileNode
-     * @returns
      */
-    public async openTab(fileNode: S3FileNode): Promise<void> {
+    public async openInReadMode(fileNode: S3FileNode): Promise<void> {
+        if (fileNode.file.sizeBytes! < SIZE_LIMIT) {
+            this.toPreview = fileNode.file.arn
+        }
+
         getLogger().verbose(`S3FileViewer: Retrieving and displaying file: ${fileNode.file.key}`)
         showOutputMessage(
             localize('AWS.s3.fileViewer.info.fileKey', 'Retrieving and displaying file: {0}', fileNode.file.key),
             this.outputChannel
         )
 
+        const charset = await this.getCharset(fileNode)
+
         const fileLocation = await this.getFile(fileNode)
         if (!fileLocation) {
             return
         }
         getLogger().verbose(`S3FileViewer: File from s3 or temp to be opened is: ${fileLocation}`)
+        const s3Uri = vscode.Uri.parse('s3:' + fileLocation.fsPath)
+
+        if (charset != 'UTF-8') {
+            const prompt = "Can't open this file type in read-only mode, do you want to try opening in edit?"
+            const edit = 'Open in edit mode'
+            const read = 'Try in read-only'
+            if (await showConfirmationMessage({ prompt, confirm: edit, cancel: read }, this.window)) {
+                await this.openInEditMode(fileNode)
+                return
+            }
+        }
 
-        const newTab = this.activeTabs.get(fileLocation.fsPath) ?? new S3Tab(fileLocation, fileNode)
-        await newTab.openFileInReadOnly()
+        let tab: S3Tab | undefined
+        if (fileNode.file.sizeBytes! < SIZE_LIMIT) {
+            const pathToPreview = await this.arnToFsPath(this.toPreview!)
+            if (s3Uri.fsPath !== pathToPreview) {
+                return
+            }
+            tab =
+                this.activeTabs.get(pathToPreview) ??
+                ({ fileUri: fileLocation, s3Uri, editor: undefined, s3FileNode: fileNode } as S3Tab)
+            await this.openFile(tab, tab.s3Uri, true)
+            this.toPreview = undefined
+        } else {
+            tab =
+                this.activeTabs.get(fileLocation.fsPath) ??
+                ({ fileUri: fileLocation, s3Uri, editor: undefined, s3FileNode: fileNode } as S3Tab)
```

Comment:
(also for line 183)
using `as` is risky when creating object literals. If we later add other required fields, this will be missed and will result in bugs. Is `as` really needed here?
```suggestion
                ({ fileUri: fileLocation, s3Uri, editor: undefined, s3FileNode: fileNode })
```

Reply:
That is just to declare that object as type S3Tab, using the above interface. Is there another way to do this?